src/server/ua_services_session.c

```diff
@@ -63,27 +138,81 @@ void Service_CreateSession(UA_Server *server, UA_SecureChannel *channel,
+    /*     response->responseHeader.serviceResult |= */
+    /*     UA_ByteString_copy(&channel->endpoint->endpointDescription.serverCertificate, */
+    /*                        &response->serverCertificate); */
+
+    /* Create a signed nonce */
+    response->responseHeader.serviceResult =
+        nonceAndSignCreateSessionResponse(server, channel, newSession, request, response);
+    
     /* Failure -> remove the session */
     if(response->responseHeader.serviceResult != UA_STATUSCODE_GOOD) {
-        UA_SessionManager_removeSession(&server->sessionManager,
-                                        &newSession->authenticationToken);
-         return;
+        UA_SessionManager_removeSession(&server->sessionManager, &newSession->authenticationToken);
+        return;
     }
 
     UA_LOG_DEBUG_CHANNEL(server->config.logger, channel,
            "Session " UA_PRINTF_GUID_FORMAT " created",
            UA_PRINTF_GUID_DATA(newSession->sessionId.identifier.guid));
 }
 
+static void
+checkSignature(const UA_Server *server,
+               const UA_SecureChannel *channel,
+               UA_Session *session,
+               const UA_ActivateSessionRequest *request,
+               UA_ActivateSessionResponse *response) {
+    if(channel->securityMode == UA_MESSAGESECURITYMODE_SIGN ||
+       channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT) {
+        const UA_SecurityPolicy *const securityPolicy = channel->securityPolicy;
+        const UA_ByteString *const localCertificate = &securityPolicy->localCertificate;
+
+        UA_ByteString dataToVerify;
+        UA_StatusCode retval = UA_ByteString_allocBuffer(&dataToVerify,
+                                                         localCertificate->length + session->serverNonce.length);
+        if(retval != UA_STATUSCODE_GOOD) {
+            response->responseHeader.serviceResult = retval;
+            UA_LOG_DEBUG_SESSION(server->config.logger, session,
+                                 "Failed to allocate buffer for signature verification! %#10x", retval);
+            return;
+        }
+
+        memcpy(dataToVerify.data, localCertificate->data, localCertificate->length);
+        memcpy(dataToVerify.data + localCertificate->length,
+               session->serverNonce.data, session->serverNonce.length);
+
+        retval = securityPolicy->asymmetricModule.signingModule.
+            verify(securityPolicy, channel->channelContext, &dataToVerify,
+                   &request->clientSignature.signature);
+        if(retval != UA_STATUSCODE_GOOD) {
+            response->responseHeader.serviceResult = retval;
+            UA_LOG_DEBUG_SESSION(server->config.logger, session,
+                                 "Failed to verify the client signature! %#10x", retval);
+            UA_ByteString_deleteMembers(&dataToVerify);
+            return;
+        }
+
+        retval  = UA_SecureChannel_generateNonce(channel, 32, &response->serverNonce);
```

Comment:
I believe it might be necessary to move this into the activateSession function so that the nonce is always regenerated, regardless of the verification status of the signature.
For example if the client tries to activate a session and it fails because of an invalid signature, and then tries to activate again, the nonce should be different according to the standard.